src/actions/dropbox/dropbox.ts

```diff
@@ -0,0 +1,187 @@
+    form.fields = []
+
+    let accessToken = ""
+    if (request.params.state_json) {
+      try {
+        const jsonState = JSON.parse(request.params.state_json)
+        if (jsonState.code && jsonState.redirect) {
+          accessToken = await this.getAccessTokenFromCode(request)
+        }
+      } catch { winston.warn("Could not parse state_json") }
+    }
+    const drop = this.dropboxClientFromRequest(request, accessToken)
+    try {
+      const response = await drop.filesListFolder({path: ""})
+      let folderList = response.entries.filter((entries) => (entries[".tag"] === "folder"))
+        .map((entries) => ({name: entries.name, label: entries.name}))
+      if (folderList.length === 0) {
+        folderList = [{name: "Looker", label: "Looker"}]
+      }
+      form.fields = [{
+        description: "Dropbox directory where file will be saved",
+        label: "Save in",
+        name: "directory",
+        options: folderList,
+        required: true,
+        type: "select",
+      }, {
+        label: "Filename",
+        name: "filename",
+        type: "string",
+      }]
+      if (accessToken !== "") {
+        const newState = JSON.stringify({access_token: accessToken})
+        form.state = new Hub.ActionState()
+        form.state.data = newState
+      }
+      return form
+    } catch (_error) {
+      const actionCrypto = new Hub.ActionCrypto()
+      const jsonString = JSON.stringify({stateurl: request.params.state_url})
+      const ciphertextBlob = await actionCrypto.encrypt(jsonString).catch((err: string) => {
+        winston.error("Encryption not correctly configured")
+        throw err
+      })
+      form.state = new Hub.ActionState()
+      form.fields.push({
+        name: "login",
+        type: "oauth_link",
+        label: "Log in with Dropbox",
+        oauth_url: `${process.env.ACTION_HUB_BASE_URL}/actions/dropbox/oauth?state=${ciphertextBlob}`,
+      })
+      return(form)
+    }
+  }
+
+  async oauthUrl(redirectUri: string, encryptedState: string) {
+    const url = new URL("https://www.dropbox.com/oauth2/authorize")
+    url.search = querystring.stringify({
+      response_type: "code",
+      client_id: process.env.DROPBOX_ACTION_APP_KEY,
+      redirect_uri: redirectUri,
+      force_reapprove: true,
+      state: encryptedState,
+    })
+    return url.toString()
+  }
+
+  async oauthFetchInfo(urlParams: { [key: string]: string }, redirectUri: string) {
+    const actionCrypto = new Hub.ActionCrypto()
+    const plaintext = await actionCrypto.decrypt(urlParams.state).catch((err: string) => {
+      winston.error("Encryption not correctly configured" + err)
+      throw err
+    })
+
+    const payload = JSON.parse(plaintext)
+    await https.post({
+      url: payload.stateurl,
+      body: JSON.stringify({code: urlParams.code, redirect: redirectUri}),
+    }).catch((_err) => { winston.error(_err.toString()) })
+    return `<html><script>window.close()</script>></html>`
```

Comment:
What's the use case for returning something other than this? Kind of annoying/scary for every action to be returning arbitrary HTML/JS. Can we standardize this?

Reply:
Need to auto close the new tab that was opened for oauth with Dropbox's authorization server. Seemed like the cleanest way to do it

Reply:
What I meant to say is can we push the scary `<html><script>window.close()</script>></html>` up into the server or base class or somewhere? Would you expect other actions to need this?